### PR TITLE
distribute framework: restrict the task state transform rule

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/41495

Problem Summary:

### What is changed and how it works?
restrict the task state transform rule
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -213,8 +216,17 @@ func (d *dispatcher) handleRunning() error {
 func (d *dispatcher) updateTask(taskState string, newSubTasks []*proto.Subtask, retryTimes int) (err error) {
 	prevState := d.task.State
 	d.task.State = taskState
+	if !VerifyTaskStateTransform(prevState, taskState) {
+		return errors.Errorf("invalid task state transform, from %s to %s", prevState, taskState)
+	}
+	failpoint.Inject("syncInUpdateTask", func(_ failpoint.Value) {
```
Comment (ywqzzy) [inline disttask/framework/dispatcher/dispatcher.go:222]: Seems not used.

File: disttask/framework/proto/task.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ const (
 	TaskStateCanceled      = "canceled"
 	TaskStatePausing       = "pausing"
 	TaskStatePaused        = "paused"
+	TaskStateResuming      = "resuming"
```
Comment (ywqzzy) [inline disttask/framework/proto/task.go:44]: Need more discussion to introduce a new state.
For now, LGTM.

Comment (ywqzzy) [reply disttask/framework/proto/task.go:44]: Further more, we can distinguish subtask state and task state later.

**Review by ywqzzy — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

File: disttask/framework/storage/task_table.go [hunk 3]
```diff
@@ -468,10 +468,10 @@ func (stm *TaskManager) GetSchedulerIDsByTaskID(taskID int64) ([]string, error)
 }
 
 // UpdateGlobalTaskAndAddSubTasks update the global task and add new subtasks
-func (stm *TaskManager) UpdateGlobalTaskAndAddSubTasks(gTask *proto.Task, subtasks []*proto.Subtask) error {
+func (stm *TaskManager) UpdateGlobalTaskAndAddSubTasks(gTask *proto.Task, subtasks []*proto.Subtask, prevState string) error {
 	return stm.WithNewTxn(stm.ctx, func(se sessionctx.Context) error {
-		_, err := ExecSQL(stm.ctx, se, "update mysql.tidb_global_task set state = %?, dispatcher_id = %?, step = %?, state_update_time = %?, concurrency = %?, meta = %?, error = %? where id = %?",
-			gTask.State, gTask.DispatcherID, gTask.Step, gTask.StateUpdateTime.UTC().String(), gTask.Concurrency, gTask.Meta, serializeErr(gTask.Error), gTask.ID)
+		_, err := ExecSQL(stm.ctx, se, "update mysql.tidb_global_task set state = %?, dispatcher_id = %?, step = %?, state_update_time = %?, concurrency = %?, meta = %?, error = %? where id = %? and state = %?",
+			gTask.State, gTask.DispatcherID, gTask.Step, gTask.StateUpdateTime.UTC().String(), gTask.Concurrency, gTask.Meta, serializeErr(gTask.Error), gTask.ID, prevState)
```
Comment (D3Hunter) [inline disttask/framework/storage/task_table.go:474]: we need check how many lines're affected, if 0, we shouldn't update subtasks

Comment (ywqzzy) [reply disttask/framework/storage/task_table.go:474]: > we need check how many lines're affected, if 0, we shouldn't update subtasks

I will fix it.

**Pull request merged**